feat(store): expose StateContextFactory, StateFactory

splincode opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/ngxs/store/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->
```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Other... Please describe:
```

## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: https://github.com/ngxs/store/issues/1323

## What is the new behavior?

For the plugin `@ngxs-labs/data` needs a more api from `@ngxs/store`. Because `@ngxs-labs/data` use deep internal manipulation with Store.

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

splincode commented:
@markwhitfeld Please, I need this functionality in production right now. Therefore, I want to use this functionality in our projects.

Now I can’t wait, so on the prod I will use

```json
{
  "@ngxs/store": "3.5.1@dev",
  "@ngxs-labs/data": "1.0.1"
}
```

This functionality is really necessary for the plugin.

splincode commented:
One of the big problems is that ngcc generates randomly `ɵm, ɵp, ɵt, ...`

![image](https://user-images.githubusercontent.com/12021443/65493853-6880cc00-debc-11e9-910d-cdec2ba9e937.png)

Therefore, the plugin does not work

splincode commented:
@markwhitfeld updated

![image](https://user-images.githubusercontent.com/12021443/65820759-97d56700-e235-11e9-86b1-5fca71c4cf87.png)

markwhitfeld commented on packages/store/src/module.ts:
```diff
@@ -132,6 +139,14 @@ export class NgxsModule {
+      },
+      {
+        provide: NGXS_INTERNAL_CONTEXT_FACTORY,
+        useValue: StateContextFactory
```
```suggestion
        useClass: StateContextFactory
```

markwhitfeld commented on packages/store/src/module.ts:
```diff
@@ -132,6 +139,14 @@ export class NgxsModule {
+      },
+      {
+        provide: NGXS_INTERNAL_FACTORY,
+        useValue: StateFactory
```
```suggestion
        useClass: StateFactory
```

markwhitfeld commented on packages/store/internals/src/internal-tokens.ts:
```diff
@@ -0,0 +1,15 @@
+/**
+ * @see only use to access internal {@link StateContextFactory} in plugins
+ */
+export const NGXS_INTERNAL_CONTEXT_FACTORY: InjectionToken<string> = new InjectionToken(
```
```suggestion
export const NGXS_INTERNAL_CONTEXT_FACTORY_TOKEN: InjectionToken<string> = new InjectionToken(
```

markwhitfeld commented on packages/store/internals/src/internal-tokens.ts:
```diff
@@ -0,0 +1,15 @@
+/**
+ * @see only use to access internal {@link StateFactory} in plugins
+ */
+export const NGXS_INTERNAL_FACTORY: InjectionToken<string> = new InjectionToken(
```
```suggestion
export const NGXS_INTERNAL_FACTORY_TOKEN: InjectionToken<string> = new InjectionToken(
```

splincode commented:
@markwhitfeld done

arturovt commented on packages/store/internals/src/internal-tokens.ts:
```diff
@@ -0,0 +1,15 @@
+/**
+ * @see only use to access internal {@link StateContextFactory} in plugins
+ */
+export const NGXS_INTERNAL_CONTEXT_FACTORY_TOKEN: InjectionToken<string> = new InjectionToken(
```
Not sure about conventions. But see as Angular team does it https://github.com/angular/angular/blob/f8096d499324cf0961f092944bbaedd05364eea1/packages/platform-webworker/src/web_workers/shared/api.ts#L11. They have `TOKEN_NAME = new InjectionToken('tokenName')`. So we could also have `NGXS_STATE_CONTEXT_FACTORY = new Token('Internals.StateContextFactory')`. 

This will ensure consistency. Also we do not come up with our own approach and do as framework creators.

splincode replied:
@markwhitfeld what do you think? I did according to the example

![image](https://user-images.githubusercontent.com/12021443/65946688-5ca68400-e43f-11e9-8cc3-a835338753b7.png)

arturovt replied:
We cannot change those tokens descriptions, but we can start doing correctly in the new package (by “new” I mean that the internals package doesn’t have any tokens yet).

markwhitfeld replied:
It looks like none of the tokens with the `_TOKEN` suffix are currently exported so we are safe to change our conventions (the others can be fixed in another PR).
@splincode could you change these new token to follow the Angular convention? Sorry to mess you around like this

splincode replied:
done

arturovt commented on packages/store/internals/src/internal-tokens.ts:
```diff
@@ -0,0 +1,15 @@
+);
+
+/**
+ * @see only use to access internal {@link StateFactory} in plugins
```
```suggestion
 * @see {@link StateFactory} as it's referenced by this token to be accessed by plugins internally
```

arturovt commented on packages/store/internals/src/internal-tokens.ts:
```diff
@@ -0,0 +1,15 @@
+import { InjectionToken } from '@angular/core';
+
+/**
+ * @see only use to access internal {@link StateContextFactory} in plugins
```
```suggestion
 * @see {@link StateContextFactory} as it's referenced by this token to be accessed by plugins internally
```

arturovt commented on packages/store/internals/src/internal-tokens.ts:
```diff
@@ -0,0 +1,15 @@
+/**
+ * @see only use to access internal {@link StateFactory} in plugins
+ */
+export const NGXS_INTERNAL_FACTORY_TOKEN: InjectionToken<string> = new InjectionToken(
```
So as we decided to change it according the Angular convention:
```suggestion
export const NGXS_STATE_FACTORY = new InjectionToken<unknown>('Internals.StateFactory')
```
And I guess it's not `<string>`, let's leave `unknown` here as it's realistically unknown (because this package doesn't know anything about the real class it references, fix me if I'm wrong).

arturovt commented on packages/store/internals/src/internal-tokens.ts:
```diff
@@ -0,0 +1,15 @@
+/**
+ * @see only use to access internal {@link StateContextFactory} in plugins
+ */
+export const NGXS_INTERNAL_CONTEXT_FACTORY_TOKEN: InjectionToken<string> = new InjectionToken(
```
```suggestion
export const NGXS_STATE_CONTEXT_FACTORY = new InjectionToken<unknown>('Internals.StateContextFactory')
```

splincode commented:
![image](https://user-images.githubusercontent.com/12021443/65993972-ff402080-e49a-11e9-92ac-9a9cee6859ce.png)

* updated
* removed link operator

arturovt commented:
Sure, I’m ok with that.

splincode commented:
@arturovt only in V4
